### ISSUE TITLE
ENH: Construct pandas dataframe from function

### ISSUE THREAD

**Issue by giuliobeseghi** [pandas-dev/pandas#33894]
#### Is your feature request related to a problem?

It would be great if we could construct a `pd.DataFrame` from a function.

#### Describe the solution you'd like

```python
import pandas as pd
import numpy as np

def my_random():
    return np.random.rand() + 1

df = pd.DataFrame(my_random, index=range(10), columns=range(10))
```

Alternatively, we could do the same with something like

```python
df = pd.DataFrame.from_function(my_random, *args, **kwargs)
```

`args` and `kwargs` are passed to the df constructor.

**Comment by MarcoGorelli** [comment]
Thanks @giuliobeseghi for the suggestion

TBH I don't see how the example you've given is different to initalising a DataFrame with a numpy array. In general, I don't see this could work with arbitrary functions. Could you please provide some more details?

**Comment by TomAugspurger** [comment]
How is this an improvement over having the function return a DataFrame? I don't see why it would need to be a classmethod.

**Comment by giuliobeseghi** [comment]
> How is this an improvement over having the function return a DataFrame? I don't see why it would need to be a classmethod.

My example was just to explain that it should work with a general function. The main reason why I proposed it is that I'd like to create dataframes like this:

`df = pd.DataFrame(np.random.rand, index=range(10), columns=range(15))`

using a function to create the values of a DataFrame. It looks clean and easy to read, but if you don't think it's an improvement we can forget about it. I thought about it because many pandas methods accept functions as arguments (I'm thinking of `loc` and `rename` for example).

**Comment by MarcoGorelli** [comment]
For your particular use-case, can't you just pass the dimensions directly to `random.rand`?
```python
pd.DataFrame(np.random.rand(15, 10)+1)
```

**Comment by giuliobeseghi** [comment]
> For your particular use-case, can't you just pass the dimensions directly to `random.rand`?
> 
> ```python
> pd.DataFrame(np.random.rand(15, 10)+1)
> ```
> 
> ## EDIT
> OK, I see what you mean now, you'd want to be able to do this even if you have some arbitrary function which doesn't take dimension arguments

Yeah, imagine for example that the index is `pd.date_range('2020', '2021', freq='45T')`. 

I see that the use case for my request is quite specific, implementing makes sense just if it's easy to carry out. It's more about style than performance.

**Comment by MarcoGorelli** [comment]
> implementing makes sense just if it's easy to carry out

That, and if it's worth maintaining + expanding the (already huge) API. I may be wrong, but I'm 99% certain the core team would be -1 on this

**Comment by jreback** [comment]
-1 on this extension, doesn't provide any new capability and adds to api bloat.

**Comment by giuliobeseghi** [comment]
It hurts... I'm joking, I understand :) I gave it a try though!

**Comment by MarcoGorelli** [comment]
It's OK :) There's lot's of other issues open if you want to contribute, see [here](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html) for how to get started
